### PR TITLE
Add Fractal RGB matrix effects

### PR DESCRIPTION
**Pull request by filterpaper**
## Description

Here is a Fractal RGB matrix effect that generates randomly lit keys on middle columns and pulse outwards, creating a horizontal fractal pattern. Animation speed is user adjustable.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/rgb_matrix_animations/pixelation_anim.h
Comment (filterpaper) [inline quantum/rgb_matrix_animations/pixelation_anim.h:23]: `rand16seed` is initialised inside `lib/lib8tion/lib8tion.c`, but that file is not included in the compile process, triggering errors when `random8()` is used. Initialising here is a workaround, but it will conflict with other similar animation code doing the same thing.

Open to better suggestion on using using `random8()` from `lib8tion` propperly.

**Comment by filterpaper** [comment]
QMK does not compile cleanly with `random8()`. Required `rand16seed` inside `random8.h` is an external variable declaration initialised in `lib8tion.c`; and compile process is unable to resolve the external reference.

A custom pseudorandom number generator, `prng8()` included instead. It is smaller than `rand()` and `random8()` because of `uint8_t` state registers. And also faster with only bit-shift and OR operators compared to multiplication used by `random8()`.

File: quantum/rgb_matrix_animations/pixelation_anim.h
Comment (filterpaper) [reply quantum/rgb_matrix_animations/pixelation_anim.h]: Will be solved by #12699

**Comment by filterpaper** [comment]
The use of `random8()` in this PR has dependency on #12699

**Review by drashna — APPROVED** [review]

**Comment by filterpaper** [comment]
Rebased against `develop` branch.

**Comment by drashna** [comment]
These need to be moved to `quantum/rgb_matrix/animations`

**Comment by filterpaper** [comment]
Update animation header file for explicit enable instead of disable. Pend this PR to merge after #13758 to align documentation text.

**Comment by drashna** [comment]
This seems to lock up on resume.

**Comment by filterpaper** [comment]
> This seems to lock up on resume.

Do you mean suspending the PC and resuming? I've not been able to reproduce on a Mac, except that the RGB may not fully turn off most of the time. Can you confirm the process to reproduce what you experienced?

**Comment by fauxpark** [comment]
> I've not been able to reproduce on a Mac, except that the RGB may not fully turn off most of the time.

You have to wait about a minute for macOS to suspend USB devices.

**Comment by filterpaper** [comment]
> > I've not been able to reproduce on a Mac, except that the RGB may not fully turn off most of the time.
> 
> You have to wait about a minute for macOS to suspend USB devices after going to sleep.

RIght, I'm aware of this because my mouse indicator LED will turn off when USB is truly suspended. However I can't reproduce the issue. I will give this a try on Windows to see if it makes a difference. Do you see the lock up problem on all three animation?

**Comment by drashna** [comment]
this is happening on Windows, for me,  but with the moonlander, specifically.   I'm going to investigate some more

**Review by tzarc — CHANGES_REQUESTED** [review]

File: docs/feature_rgb_matrix.md
Comment (filterpaper) [reply docs/feature_rgb_matrix.md:482]: @drashna Do I need to wait for #13758 to be merged, and rebase mine branch to change docs to "ENABLE"?

Comment (drashna) [reply docs/feature_rgb_matrix.md:482]: I'd say revert the change for the `ENABLE` part.  If one or the other lands first, then it would need to be updated.

**Comment by drashna** [comment]
> > > I've not been able to reproduce on a Mac, except that the RGB may not fully turn off most of the time.
> > 
> > 
> > You have to wait about a minute for macOS to suspend USB devices after going to sleep.
> 
> RIght, I'm aware of this because my mouse indicator LED will turn off when USB is truly suspended. However I can't reproduce the issue. I will give this a try on Windows to see if it makes a difference. Do you see the lock up problem on all three animation?

It seems to be ... finicky about when it will happen.  It no longer happens on my moonlander, but it seems to be happening on my corne (avr), now.  But only when in the pixel animations.

**Comment by filterpaper** [comment]
If the animation bug is transient, we can skip this PR for Q3 breaking change. I'm expecting another per key RGB PCB to test this a bit more later this month.

There are 3 "pixel" named animation; which ones are causing you problems? "Pixel Rain" is a different variation of the current "Jellybean" effect. Do you have issues as well with the latter?

**Comment by drashna** [comment]
The one that hs the issue, mostly seems to be the 3rd on, the pix fractal.

**Comment by tzarc** [comment]
Fixed up the conflicts after the merge of #13758.
Deferred until next cycle.

File: docs/feature_rgb_matrix.md
Comment (fauxpark) [inline docs/feature_rgb_matrix.md:478]:
```suggestion
|`#define DISABLE_RGB_MATRIX_HUE_WAVE`                  |Disables `RGB_MATRIX_HUE_WAVE `                |
|`#define DISABLE_RGB_MATRIX_FRACTAL`                   |Disables `RGB_MATRIX_FRACTAL`                  |
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

**Review by tzarc — APPROVED** [review]

**Pull request merged**
